java/algorithm/src/main/java/org/apache/arrow/algorithm/dictionary/SearchTreeBasedDictionaryBuilder.java

```diff
@@ -0,0 +1,171 @@
+ * </p>
+ * @param <V> the dictionary vector type.
+ */
+public class SearchTreeBasedDictionaryBuilder<V extends ValueVector> {
+
+  /**
+   * The dictionary to be built.
+   */
+  private final V dictionary;
+
+  /**
+   * The criteria for sorting in the search tree.
+   */
+  protected final VectorValueComparator<V> comparator;
+
+  /**
+   * The total number of data items probed.
+   */
+  private int valuesProbed;
+
+  /**
+   * If null should be encoded.
+   */
+  private final boolean encodeNull;
+
+  /**
+   * The search tree for storing the value index.
+   */
+  private TreeSet<Integer> searchTree;
+
+  /**
+   * Construct a search tree-based dictionary builder.
+   * @param dictionary the dictionary vector.
+   * @param comparator the criteria for value equality.
+   */
+  public SearchTreeBasedDictionaryBuilder(V dictionary, VectorValueComparator<V> comparator) {
+    this(dictionary, comparator, false);
+  }
+
+  /**
+   * Construct a search tree-based dictionary builder.
+   * @param dictionary the dictionary vector.
+   * @param comparator the criteria for value equality.
+   * @param encodeNull if null values should be added to the dictionary.
+   */
+  public SearchTreeBasedDictionaryBuilder(V dictionary, VectorValueComparator<V> comparator, boolean encodeNull) {
+    this.dictionary = dictionary;
+    this.comparator = comparator;
+    this.encodeNull = encodeNull;
+    this.comparator.attachVector(dictionary);
+
+    searchTree = new TreeSet<>((index1, index2) -> comparator.compare(index1, index2));
+
+    valuesProbed = 0;
+  }
+
+  /**
+   * Gets the dictionary built.
+   * Please note that the dictionary is not in sorted order.
+   * Instead, its order is determined by the order of element insertion.
+   * To get the dictionary in sorted order, please use
+   * {@link SearchTreeBasedDictionaryBuilder#populateSortedDictionary(ValueVector)}.
+   * @return the dictionary.
+   */
+  public V getDictionary() {
+    return dictionary;
+  }
+
+  /**
+   * Try to add all values from the target vector to the dictionary.
+   * @param targetVector the target vector containing values to probe.
+   * @return the number of values actually added to the dictionary.
+   */
+  public int addValues(V targetVector) {
+    int ret = 0;
+    for (int i = 0; i < targetVector.getValueCount(); i++) {
+      if (!encodeNull && targetVector.isNull(i)) {
+        continue;
+      }
+      if (addValue(targetVector, i)) {
+        dictionary.setValueCount(dictionary.getValueCount() + 1);
+        ret += 1;
+      }
+    }
+    return ret;
+  }
+
+  /**
+   * Try to add an element from the target vector to the dictionary.
+   * @param targetVector the target vector containing new element.
+   * @param targetIndex the index of the new element in the target vector.
+   * @return true if the element is added to the dictionary, and false otherwise.
+   */
+  public boolean addValue(V targetVector, int targetIndex) {
+    valuesProbed += 1;
```

Comment:
what is this?

Reply:
Basic stats, not used for now. So I will remove it.